### PR TITLE
[Feature](mluOpRoiAlignRotatedForward): Refactor roi align rotated forward

### PR DESCRIPTION
**Pull request by chqy99**
Thanks for your contribution and we appreciate it a lot. :rocket::rocket:

## 1. Motivation

Please describe your motivation and the goal you want to achieve through this pull request.

## 2. Modification

Please briefly describe what modification is made in this pull request, and indicate where to make the modification.

Are new test cases added? If so, please post the corresponding generator-PR link here.

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

### CODE REVIEW

**Comment by mahxn0** [comment]
1. 4次gather，每次gather一个点 可以修改为一个gather 4个点，看看是否有性能优化

**Comment by mahxn0** [comment]
2. 坐标计算部分向量化实现下

File: kernels/roi_align_rotated/roi_align_rotated_forward_vector.mlu [hunk 1]
```diff
@@ -0,0 +1,585 @@
...
+    ((uint32_t *)w3)[i] = x_high;
+    pos1[i] = y_low;
+    pos3[i] = y_high;
+
+    // nram_y cache ly, nram_x cache lx
+    // T ly = y - y_low;
+    // T lx = x - x_low;
+    nram_y[i] = y - y_low;
+    nram_x[i] = x - x_low;
+  }
```
Comment (mahxn0) [inline kernels/roi_align_rotated/roi_align_rotated_forward_vector.mlu:236]: 194-236能否向量化实现  参考ms_deform

**Comment by chqy99** [comment]
> 1. 4次gather，每次gather一个点 可以修改为一次gather 4个点，看看是否有性能优化

对于数据量较小的情况性能提升明显，10 % ~ 20 %

File: kernels/roi_align_rotated/roi_align_rotated_forward_vector.mlu (hunk 1 above)
Comment (chqy99) [reply kernels/roi_align_rotated/roi_align_rotated_forward_vector.mlu:236]: 这个类型不一样，空间不够用了

File: kernels/roi_align_rotated/roi_align_rotated_forward_vector.mlu [hunk 2]
```diff
@@ -0,0 +1,585 @@
...
+    pos_offset = align_offset;
+  }
+  uint32_t hwc_num = deal_channels * vec_num;
+
+  __gather(val, input, pos, deal_channels * sizeof(T), GDRAM2NRAM,
+           deal_channels * sizeof(T), vec_num);
+  if (deal_channels != 1) {
+    __bang_transpose(v_t, val, vec_num, deal_channels);
+    __bang_cycle_mul(v_t, v_t, w, hwc_num, vec_num);
+    __bang_transpose(val, v_t, deal_channels, vec_num);
```
Comment (mahxn0) [inline kernels/roi_align_rotated/roi_align_rotated_forward_vector.mlu:326]: 可以先trans w吗

**Review by mahxn0 — APPROVED** [review]

File: kernels/roi_align_rotated/roi_align_rotated_forward_vector.mlu (hunk 2 above)
Comment (chqy99) [reply kernels/roi_align_rotated/roi_align_rotated_forward_vector.mlu:326]: 扩展 w 需要额外空间，之前测试过没有收益

**Review by DanieeelLiu — APPROVED** [review]
LGTM

**Review by mahxn0 — APPROVED** [review]

**Comment by mahxn0** [comment]
LGTM

**Review by duzekunKTH — APPROVED** [review]

**Pull request merged**